scipy/stats/kde.py

```diff
@@ -619,6 +625,132 @@ def logpdf(self, x):
+        
+        """
+        points = atleast_2d(x)
+        ax = atleast_1d(axis)
+
+        if not issubclass(ax.dtype.type, numbers.Integral):
+            msg = "elements of axis should be intergers"
+            raise ValueError(msg)
+
+        d, m = points.shape
+        if ax.size > self.d or ax.size == 0:
+            msg = "axis have dimension %s, dataset has dimension %s" % \
+                (ax.size, self.d)
+            raise ValueError(msg)
+
+        if d != ax.size:
+            if d == 1 and m == ax.size:
+                # points was passed in as a row vector
+                points = reshape(points, (m, 1))
+                d = ax.size
+                m = 1
+            else:
+                msg = "points have dimension %s, axis has dimension %s" % \
+                (d, ax.size)
+
+        cov = self.covariance[ax[:, newaxis], ax]
+        inv_cov = linalg.inv(cov)
+        
+        whitening = linalg.cholesky(inv_cov).T
+        scaled_dataset = dot(whitening, self.dataset[ax])
+        scaled_points = dot(whitening, points)
+        norm_factor = sqrt(linalg.det(2*pi*cov))
+
+        if m >= self.n:
+            # there are more points than data, so loop over data
+            energy = np.array([sum((scaled_dataset[:, i, newaxis] - scaled_points)
+                * (scaled_dataset[:, i, newaxis] - scaled_points), axis=0) / 2.0 
+                for i in range(self.n)])
+            result = dot(self.weights, exp(-energy))
+        else:
+            # loop over points
+            result = np.array([sum(exp(-sum((scaled_dataset - 
+                scaled_points[:, i, newaxis]) * (scaled_dataset - 
+                scaled_points[:, i, newaxis]), axis=0) / 2.0) * self.weights, 
+                axis=0) for i in range(m)])
+
+        result = result / norm_factor
+
+        return result
+
+    def logpdf_marginal(self, x, axis):
+        """
+        Evaluate the log of estimated marginal pdf on a provided set of points.
+        
+        Notes
+        -----
+        See the ``pdf_marginal`` docstring for more details.
+        
+        """
+        points = atleast_2d(x)
+        ax = atleast_1d(axis)
+
+        if not issubclass(ax.dtype.type, numbers.Integral):
+            msg = "elements of axis should be intergers"
+            raise ValueError(msg)
+
+        d, m = points.shape
+        if ax.size > self.d or ax.size == 0:
+            msg = "axis have dimension %s, dataset has dimension %s" % \
+                (ax.size, self.d)
+            raise ValueError(msg)
+
+        if d != ax.size:
+            if d == 1 and m == ax.size:
+                # points was passed in as a row vector
+                points = reshape(points, (m, 1))
+                d = ax.size
+                m = 1
+            else:
+                msg = "points have dimension %s, axis has dimension %s" % \
+                (d, ax.size)
+
+        cov = self.covariance[ax[:, newaxis], ax]
+        inv_cov = linalg.inv(cov)
```

Comment:
Same as above.

Reply:
Ah yes I was too fast. Sorry about that. Then you can just use `solve(cov, dataset[:, i, newaxis] - points)`. In general, you can get away without inverting anything if the inverse appears in an algebraic expression. Obviously, the determinant then stays as is this time.

Reply:
Sure.